Parser: Allow escaped quotes in attribute strings

dmsnell opened the pull request:
Fixes #1086 
Depends on #1142 

Previously the parser would fail handling attributes with escaped quotes
in them. For example, 'data="a\"a"' would fail because the parser
attempted to grab only `a\` instead of `a\"a`.

This has been fixed by indicating that a valid value is _either_ an
escaped quote _or_ any character other than a quote.

@westonruter I put this together before I saw #1088. Are we sure we want to include a full JSON parser in this as well? That may be fine, but I think it's also valid to grab the string and let JavaScript do the JSON parsing (this is how I set it up in [wp-post-grammar](https://github.com/Automattic/wp-post-grammar/blob/dac539b126df6012de6a8a7c3b70b9ad8988f514/src/post.pegjs#L1)) Doing it in JS would be faster but obviously make it less viable for harmony without effort on the PHP side.

dmsnell committed:
Parser: Accept JSON-encoded attributes

Previously the parser would fail handling attributes with escaped quotes
in them. For example, 'data="a\"a"' would fail because the parser
attempted to grab only `a\` instead of `a\"a`.

This has been fixed by indicating that a valid value is _either_ an
escaped quote _or_ any character other than a quote.

dmsnell commented:
@youknowriad hope I'm not stepping on your toes but I was curious about the reducer in the grammar that it looks like you added. I updated to the declarative version from wp-post-grammar. Was there a reason you didn't want to use list destructuring?

dmsnell commented:
> Was there a reason you didn't want to use list destructuring?

actually, is this why the tests are failing? `UglifyJS` error…

dmsnell committed:
Use objects instead of list pairs for attributes

dmsnell committed:
Use "function" syntax

dmsnell committed:
Use helper function due to missing syntax support

mcsf reviewed (approved):
As discussed in person, this exposes some _interesting_ behavior with single and double quotes — as experienced by switching between Visual and Text and doing a before-and-after comparison of which quotes are used where. Nevertheless, those aren't concerns of the parser and they should be addressed in another PR.

dmsnell committed:
Remove single-quoted attributes

WordPress is changing single-quotes to double-quotes. In simple
attributes this is fine but with complex attributes having
possibly-escaped double-quotes it will break the parser.

The simple solution is to require only double-quoted attributes with
proper escaping.

nb commented:
Do we even need the `data=`? Why not just `<!-- wp:baba {key: [{deep: value}]} -->`?

youknowriad commented:
The title of the PR is a bit misleading here. I guess this PR addresses the possibility to have escaped `"` in block values. which LGTM 👍 .

If we want to JSON representation of attributes instead of HTML-attributes, we should go the way proposed in #1088 (and drop support of the HTML attributes) or extracts an "attribute string" from the grammar and use JSON.parse globally for each block.

youknowriad commented:
Also, I guess we should also update the serializer to escape quotes in attributes.

dmsnell commented:
> Do we even need the data=? Why not just <!-- wp:baba {key: [{deep: value}]} -->?

@nb I feel like we could make that decision but it hasn't been made yet whereas the prevailing choice has been to have named parameters. in my own musing I also felt like we could explore a special attribute whose meaning is JSON data, such as `json=` or probably better `data=`. Though I'm reluctant to make a special exception for it.

As far as I can tell, and @westonruter brought this up to, we have certain ambiguity when we treat values as JSON without knowing that they are, while there are reasonable use cases for not having JSON data. Non-string and non-number data types are going to give us trouble as it's impossible to know if `value="null"` is `null` or `"null"`.

if we were to add some top-level JSON in the comment I would think having a tag would be useful. Maybe something like `json:{"key": [{"deep": "value"}]}`. I'm also noting that _JSON_ requires quotes around key names while POJOs (plain-old-JavaScript-objects) don't.

dmsnell committed:
Add back single-quoted attributes

dmsnell commented:
> Also, I guess we should also update the serializer to escape quotes in attributes.

@youknowriad added in #1142

nb commented:
Thanks for mentioning, #1088, @youknowriad, I left a supportive comment there :)

@dmsnell totally agree with your points, I think moving all the way to JSON will solve them all.

youknowriad reviewed (approved):
Current parser is broken, so I'd suggest merging this and continue the JSON discussion in #1088

dmsnell commented:
Went ahead and merged without adding tests because I was struggling to get the tests to work - seemed like something was mangling the quotes/escaped-quotes in my test function. It was fixing the broken behavior in manual tests. Hopefully we can add tests later.

nylen commented:
> Went ahead and merged without adding tests because I was struggling to get the tests to work - seemed like something was mangling the quotes/escaped-quotes in my test function.

This is a strange decision to me.  At the very least a follow-up issue to add tests should have been created before merging this PR.

dmsnell commented:
> This is a strange decision to me. At the very least a follow-up issue to add tests should have been created before merging this PR.

You are right. I merged because the parser was broken and now it's less broken but I should have added an issue with it. Someone on Slack did say they were working on tests and I didn't understand them. Thanks for the poke @nylen - I :heart: tests too!

BE-Webdesign commented:
>Someone on Slack did say they were working on tests and I didn't understand them

If that was me, always feel free to ask for clarification. I am pretty sure the parser does handle escaped quotes correctly.  We still should add those test cases in.

dmsnell commented:
@BE-Webdesign actually I'm pretty sure I found a bug. Also @westonruter pointed out to me today that we need much better handling.

Honestly I'm confused at how all these slashes and quotes are doing. I want to move everything over to using actual character codes `\u{53}` and `\u{22}` for example because it gets confusing fast when escaping in source code in a JSON string in a JS file…it was at this point I vented on Slack and you came to the rescue :wink:

Honestly, I think this is all good reason to consider moving the parser back out of the project. Being in a separate repository is going to mean we have to version-bump it here whenever we do updates, but I think having its own place and its own tests and its own exploratory environment can be really good. That's a tangent though…
